### ISSUE TITLE
Allow to change the position of join expressions

### ISSUE THREAD

**Issue by TheNoim** [typeorm/typeorm#6456]
**Issue type:**

[ ] question
[ ] bug report
[X] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[X] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[X] `0.2.25` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Currently, there is no way to change the position of generated join expression from typeorm. The issue I have with this is that maybe you want to access the alias of a join expression from another from expression you added by yourself. 

This is what typeorm currently does (simplified):

```sql
select
	/* Selects */
from
	"job" "job",
	/* from expressions added by me */
	/* want to access the company alias. But I can't, because the left join is add the end. */
	setweight(to_tsvector('german', coalesce(company.name, '')), 'A') "textsearch_2"
/* Added by typeorm */	
left join "company" "company" on
	"company"."id" = "job"."companyId"
/* more joins */
where
	/* wheres */
```

This does not work. You can not access the company alias, because the left join was placed at the end.

It can work with this:

```sql
select
	/* Selects */
from
	"job" "job"
/* Add the join here */	
left join "company" "company" on
	"company"."id" = "job"."companyId",
	/* from expressions added by me */
	/* Now I cann access the company alias. */
	setweight(to_tsvector('german', coalesce(company.name, '')), 'A') "textsearch_2"
where
	/* wheres */
```

I already implemented exactly the feature I need for this:

https://github.com/TheNoim/typeorm/commit/43b32a92a0c2904c99cd17642b7c1ad95b68bda2

A new joinAfterIndex query call. With it, you can move the join expression to a different position. 

For example, I simply can do this `query.joinAfterIndex(0)` to get the query I described above.

I am not sure if such a change would be welcome. That's why I open an issue first before I implement tests and do a pull request.

**Comment by imnotjames** [comment]
Do you have any examples of where else this feature exists?

**Comment by imnotjames** [comment]
Actually. Looking at your example... That's valid? 🤔

**Comment by TheNoim** [comment]
Yes, it is. Maybe I missed something because of the simplification. But overall, yes it is valid, and I use it in production. 

I think [my current](https://github.com/TheNoim/typeorm/commit/43b32a92a0c2904c99cd17642b7c1ad95b68bda2) solution isn't really something I would try to push into typeorm. There needs to be a more elegant solution. The `joinAfterIndex()` method was more of an emergency thing.

Maybe add an index parameter to the join operation. Example:

```typescript
leftJoinAndSelect(tableName: string, alias: string, condition?: string, parameters?: ObjectLiteral, joinPosition?: number): this;
```
